lib/BaseDownloadController.js

```diff
@@ -157,6 +173,50 @@ module.exports = class BaseDownloadController extends CompositeController {
     }
   }
 
+  async _fetchHeaderSecrets(requestOptions) {
+    requestOptions = clone(requestOptions);
+    let headers = objectPath.get(requestOptions, 'headers');
+    if (headers) {
+      for (let [hKey, hObject] of Object.entries(headers)) {
+        let secretRef = objectPath.get(hObject, 'valueFrom.secretKeyRef');
+        if (secretRef) {
+          let secretName = objectPath.get(secretRef, 'name');
+          let secretNamespace = objectPath.get(secretRef, 'namespace', this.namespace);
+          let secretKey = objectPath.get(secretRef, 'key');
+          try {
+            objectPath.set(headers, [hKey], await this._getSecretData(secretName, secretKey, secretNamespace));
+          } catch (e) {
+            throw Error(`Unable to fetch header secret data. { name: ${secretName}, namespace: ${secretNamespace }, key: ${secretKey} }: ${objectPath.get(e, 'error.message')}`);
+          }
+        }
+      }
+    }
+    return requestOptions;
+  }
+
+  async _getSecretData(name, key, ns) {
+    let res = await this.kubeResourceMeta.request({ uri: `/api/v1/namespaces/${ns || this.namespace}/secrets/${name}`, json: true });
+    let secret = Buffer.from(objectPath.get(res, ['data', key], ''), 'base64').toString();
+    if (secret === '') {
+      throw {
+        name: 'StatusCodeError',
+        statusCode: 404,
+        message: `404 - key "${key}" not found in secret "${name}"`,
+        error: {
+          kind: 'Status',
+          apiVersion: 'v1',
+          metadata: {},
+          status: 'Failure',
+          message: `key "${key}" not found in secret "${name}"`,
```

Comment:
Suggest include the secret namespace as well

Reply:
updated